{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from datetime import date, datetime\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pandas\n",
    "import pandas_profiling\n",
    "import plotnine\n",
    "from plotnine import *  # Provides a ggplot-like interface to matplotlib.\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import seaborn as sns\n",
    "import math\n",
    "import datetime"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GWAS summary generation Step for AllofUS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SAIGE - Step1 - GRM and Null Model Fitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "traits = pd.read_csv('./Phenotypes/eur_basic_traits.tsv', sep = '\\t')\n",
    "traits = traits.columns[15:124]\n",
    "num_traits = len(traits)\n",
    "print(num_traits)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket = os.getenv('WORKSPACE_BUCKET')\n",
    "gwas_dir = 'SAIGE_GENE'\n",
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')\n",
    "%env USER_NAME={USER_NAME}\n",
    "JOB_NAME=f'saige-step1-{USER_NAME}'\n",
    "%env JOB_NAME={JOB_NAME}\n",
    "%env PHEN={phen_name}\n",
    "\n",
    "params_df = pd.DataFrame(data={\n",
    "    '--input-recursive INPUT_DIR': [f\"{bucket}/{gwas_dir}/step1_input/\"]*num_traits,\n",
    "    '--input-recursive GRM_DIR': [f\"{bucket}/{gwas_dir}/step0_output/\"]*num_traits,\n",
    "    '--output-recursive OUT_DIR': [f\"{bucket}/{gwas_dir}/step1_output/\"]*num_traits,\n",
    "    '--env traitType':['quantitative' if x.__contains__('f.') else 'binary' for x in traits],\n",
    "    '--env invNormalize':[True if x.__contains__('f.') else False for x in traits],    \n",
    "    '--env PHEN': traits\n",
    "})\n",
    "\n",
    "PARAMETER_FILENAME = f'{JOB_NAME}_params.tsv'\n",
    "%env PARAMETER_FILENAME={PARAMETER_FILENAME}\n",
    "\n",
    "params_df.to_csv(PARAMETER_FILENAME, sep='\\t', index=False)\n",
    "\n",
    "job_output = !source ~/aou_dsub.bash; aou_dsub \\\n",
    "  --name \"${JOB_NAME}_${PHEN}\" \\\n",
    "  --provider google-cls-v2 \\\n",
    "  --image \"wzhou88/saige:1.3.0\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/dsub_logs/saige_step1\" \\\n",
    "  --boot-disk-size 50 \\\n",
    "  --disk-size 128 \\\n",
    "  --min-ram 32 \\\n",
    "  --min-cores 8 \\\n",
    "  --tasks \"${PARAMETER_FILENAME}\" \\\n",
    "  --command 'step1_fitNULLGLMM.R \\\n",
    "                --sparseGRMFile=$GRM_DIR/sparseGRM_relatednessCutoff_0.125_1000_randomMarkersUsed.sparseGRM.mtx \\\n",
    "                --sparseGRMSampleIDFile=$GRM_DIR/sparseGRM_relatednessCutoff_0.125_1000_randomMarkersUsed.sparseGRM.mtx.sampleIDs.txt \\\n",
    "                --useSparseGRMtoFitNULL=TRUE \\\n",
    "                --bedFile=$INPUT_DIR/pruned_arrays_eur.bed \\\n",
    "                --bimFile=$INPUT_DIR/pruned_arrays_eur_new.bim \\\n",
    "                --famFile=$INPUT_DIR/pruned_arrays_eur.fam \\\n",
    "                --phenoFile=$INPUT_DIR/eur_basic_traits.tsv  \\\n",
    "                --invNormalize=$invNormalize \\\n",
    "                --phenoCol=${PHEN} \\\n",
    "                --covarColList=SEX,AGE,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10 \\\n",
    "                --qCovarColList=SEX \\\n",
    "                --sampleIDColinphenoFile=IID \\\n",
    "                --traitType=$traitType \\\n",
    "                --isCateVarianceRatio=TRUE \\\n",
    "                --outputPrefix=$OUT_DIR/${PHEN}_step1_output \\\n",
    "                --IsOverwriteVarianceRatioFile=TRUE \\\n",
    "                --LOCO=FALSE \\\n",
    "                --nThreads=8'\n",
    "job_output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SAIGE - Step2 - Single Variant Association Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Step 2: Single variant association testing\n",
    "\n",
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')\n",
    "%env USER_NAME={USER_NAME}\n",
    "JOB_NAME=f'saige-step2-{USER_NAME}'\n",
    "%env JOB_NAME={JOB_NAME}\n",
    "bucket = os.getenv('WORKSPACE_BUCKET')\n",
    "gwas_dir = 'SAIGE_GENE'\n",
    "\n",
    "\n",
    "exome_bgen_dir = \"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome_v7.1/bgen\"\n",
    "\n",
    "params_df = pd.DataFrame(data={\n",
    "    '--input-recursive STEP1_OUT': [f\"{bucket}/{gwas_dir}/step1_output\" for _ in range(22*num_traits)],\n",
    "    '--input bgenFile': [f\"{exome_bgen_dir}/exome.chr{x+1}.bgen\" for x in range(22)]*num_traits,\n",
    "    '--input bgenFileIndex': [f\"{exome_bgen_dir}/exome.chr{x+1}.bgen.bgi\" for x in range(22)]*num_traits,\n",
    "    '--input sampleFile': [f\"{exome_bgen_dir}/exome.chr{x+1}.sample\" for x in range(22)]*num_traits,\n",
    "    '--input groupFile': [f\"{bucket}/{gwas_dir}/step2_input/UKB470k_chr_{x+1}_groupfile.loftee.edit.txt\" for x in range(22)]*num_traits,\n",
    "    '--env PHEN': [x for x in traits for _ in range(22)],\n",
    "    '--env CHROM': [x+1 for x in range(22)]*num_traits,\n",
    "    '--output-recursive OUT_DIR': [f\"{bucket}/{gwas_dir}/step2_output/\" for _ in range(22*num_traits)]\n",
    "})\n",
    "\n",
    "\n",
    "PARAMETER_FILENAME = f'{JOB_NAME}_params.tsv'\n",
    "%env PARAMETER_FILENAME={PARAMETER_FILENAME}\n",
    "\n",
    "params_df.to_csv(PARAMETER_FILENAME, sep='\\t', index=False)\n",
    "job_id = !source ~/aou_dsub.bash; aou_dsub \\\n",
    "  --name \"${JOB_NAME}\" \\\n",
    "  --provider google-batch \\\n",
    "  --image \"wzhou88/saige:1.3.0\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/dsub_logs/saige_step2\" \\\n",
    "  --boot-disk-size 100 \\\n",
    "  --disk-size 200 \\\n",
    "  --min-ram 32 \\\n",
    "  --tasks \"${PARAMETER_FILENAME}\" \\\n",
    "  --command 'step2_SPAtests.R \\\n",
    "                --bgenFile=${bgenFile} \\\n",
    "                --bgenFileIndex=${bgenFileIndex} \\\n",
    "                --minMAF=0 \\\n",
    "                --AlleleOrder=ref-first \\\n",
    "                --is_output_moreDetails=TRUE \\\n",
    "                --is_overwrite_output=TRUE \\\n",
    "                --chrom=chr${CHROM} \\\n",
    "                --GMMATmodelFile=$STEP1_OUT/${PHEN}_step1_output.rda \\\n",
    "                --varianceRatioFile=$STEP1_OUT/${PHEN}_step1_output.varianceRatio.txt \\\n",
    "                --LOCO=FALSE \\\n",
    "                --SAIGEOutputFile=$OUT_DIR/${PHEN}_chr${CHROM}_step2_output_single.txt \\\n",
    "                '\n",
    "print(\"\\n\".join(job_id))\n",
    "job_id = job_id[1].split(\" \")[-1]\n",
    "%env JOB_ID={job_id}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sparse LD Matrix Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd_line = 'step3_LDmat.R \\\n",
    "    --bgenFile=${bgenFile} \\\n",
    "    --bgenFileIndex=${bgenFileIndex} \\\n",
    "    --sample_include_inLDMat_File=$ids \\\n",
    "    --AlleleOrder=ref-first \\\n",
    "    --chrom=$CHROM \\\n",
    "    --SAIGEOutputFile=$OUT_DIR/${maf}_${anno}_chr${CHR} \\\n",
    "    --groupFile=$groupFile/UKB470k_chr_${CHR}_groupfile.txt \\\n",
    "    --annotation_in_groupTest=${annotation_in_groupTest} \\\n",
    "    --is_overwrite_output=TRUE \\\n",
    "    --maxMAF_in_groupTest=${maf} \\\n",
    "    '\n",
    "with open(\"Cmd_step3.sh\", \"w\") as text_file:\n",
    "    text_file.write(cmd_line)\n",
    "\n",
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')\n",
    "%env USER_NAME={USER_NAME}\n",
    "JOB_NAME=f'saige-step3-{USER_NAME}'\n",
    "%env JOB_NAME={JOB_NAME}\n",
    "bucket = os.getenv('WORKSPACE_BUCKET')\n",
    "gwas_dir = 'SAIGE_GENE'\n",
    "\n",
    "exome_plink_dir = \"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome/plink_bed\"\n",
    "\n",
    "exome_bgen_dir = \"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome_v7.1/bgen\"\n",
    "anno = ['missense_lof_synonymous']\n",
    "chr = [x+1 for x in range(22)]\n",
    "mafs=([0.01)*22\n",
    "i=0\n",
    "print(f\"{bucket}/{gwas_dir}/step3_output/{mafs[i]}_{anno[i]}_chr{str(chr[i])}/\")\n",
    "params_df = pd.DataFrame(data={\n",
    "    '--input ids': [f\"{bucket}/{gwas_dir}/step1_input/ehr_IDs.txt\" for _ in range(22)],\n",
    "    '--input bgenFile': [f\"{exome_bgen_dir}/exome.chr{x+1}.bgen\" for x in range(22)],\n",
    "    '--input bgenFileIndex': [f\"{exome_bgen_dir}/exome.chr{x+1}.bgen.bgi\" for x in range(22)],\n",
    "    '--input-recursive groupFile': [f\"{bucket}/{gwas_dir}/step2_input/\" for _ in range(22)],\n",
    "    '--env maf': mafs,\n",
    "    '--env CHR': chr,\n",
    "    '--env CHROM': ['chr'+str(x+1) for x in range(22)],\n",
    "    '--output-recursive OUT_DIR': [f\"{bucket}/{gwas_dir}/step3_output\" for i in range(22)],\n",
    "    '--env annotation_in_groupTest': ['missense;lof;synonymous']*22,\n",
    "    '--env anno': anno\n",
    "})\n",
    "\n",
    "PARAMETER_FILENAME = f'{JOB_NAME}_params.tsv'\n",
    "%env PARAMETER_FILENAME={PARAMETER_FILENAME}\n",
    "\n",
    "params_df.to_csv(PARAMETER_FILENAME, sep='\\t', index=False)\n",
    "\n",
    "job_id = !source ~/aou_dsub.bash; aou_dsub \\\n",
    "  --name \"${JOB_NAME}\" \\\n",
    "  --provider google-batch \\\n",
    "  --regions us-central1 \\\n",
    "  --image \"wzhou88/saige:1.3.0\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/dsub_logs/saige_step3\" \\\n",
    "  --boot-disk-size 50 \\\n",
    "  --disk-size 100 \\\n",
    "  --min-ram 10 \\\n",
    "  --tasks \"${PARAMETER_FILENAME}\" \\\n",
    "  --script \"Cmd_step3.sh\"\n",
    "print(\"\\n\".join(job_id))\n",
    "job_id = job_id[1].split(\" \")[-1]\n",
    "%env JOB_ID={job_id}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Meta-SAIGE Analysis between AllofUS and UKB (EUR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cmd_line = '''INPUT_DIR1=${BUCKET}/SAIGE_GENE/imported/step3_docker/step3_docker/WES470k_${maf}_${anno}_chr${chrom}/\n",
    "INPUT_DIR2=${BUCKET}/SAIGE_GENE/step3_output/${maf}_${anno}_chr${chrom}/\n",
    "INPUT_GWAS1=${BUCKET}/SAIGE_GENE/imported/step2_phenome/step2/WES470k_Whites_${Phecode}_chr${chrom}_gene\n",
    "INPUT_GWAS2=${BUCKET}/SAIGE_GENE/step2_output/Pheno_${Phecode}_chr${chrom}_step2_output_single.txt\n",
    "head -1 $INPUT_GWAS1 > /app/tmp_GWAS1.txt ; tail -n+2 $INPUT_GWAS1 | awk 'BEGIN{OFS=\"\\\\t\"}$1=\"chr\"$1, $3=\"chr\"$3'  >> /app/tmp_GWAS1.txt \n",
    "head -1 $INPUT_DIR1/WES470k_chr${chrom}.marker_info.txt > /app/tmp_mkr_info1.txt ; tail -n+2 $INPUT_DIR1/WES470k_chr${chrom}.marker_info.txt | awk 'BEGIN{OFS=\"\\\\t\"}$1=\"chr\"$1' >> /app/tmp_mkr_info1.txt \n",
    "/usr/bin/time --verbose Rscript /app/RV_meta_GC.R \\\n",
    "--num_cohorts 2 \\\n",
    "--chr ${chrom} \\\n",
    "--col_co 10 \\\n",
    "--trait_type binary \\\n",
    "--verbose FALSE \\\n",
    "--info_file_path /app/tmp_mkr_info1.txt $INPUT_DIR2/${maf}_${anno}_chr${chrom}_loftee.marker_info.txt \\\n",
    "--gene_file_prefix $INPUT_DIR1/WES470k_chr${chrom}_ $INPUT_DIR2/${maf}_${anno}_chr${chrom}_loftee_ \\\n",
    "--gwas_path /app/tmp_GWAS1.txt $INPUT_GWAS2 \\\n",
    "--output_prefix $OUT_DIR/Meta_${Phecode}_${maf}_${anno}_chr${chrom}_loftee \\\n",
    "    '''\n",
    "with open(\"Cmd_meta.sh\", \"w\") as text_file:\n",
    "    text_file.write(cmd_line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')\n",
    "%env USER_NAME={USER_NAME}\n",
    "JOB_NAME=f'saige-step4-{USER_NAME}'\n",
    "%env JOB_NAME={JOB_NAME}\n",
    "bucket = os.getenv('WORKSPACE_BUCKET')\n",
    "gwas_dir = 'SAIGE_GENE'\n",
    "\n",
    "exome_bgen_dir = \"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/exome/bgen\"\n",
    "\n",
    "phecodes = pd.read_csv('phecodes.txt',header = None, dtype=object)\n",
    "phecodes = phecodes.drop_duplicates()\n",
    "phecodes = phecodes[0].tolist()\n",
    "phecodes_all = list(np.repeat(phecodes, 22*3*3))\n",
    "tot_len = len(phecodes)*3*3*22\n",
    "maf=(['0.01']*22+['0.001']*22+['0.0001']*22)*3*len(phecodes)\n",
    "anno=(['lof']*66+['missense_lof']*66+['missense_lof_synonymous']*66)*len(phecodes)\n",
    "chrom=[i for i in range(1,23)]*3*3*len(phecodes)\n",
    "\n",
    "params_df = pd.DataFrame(data={\n",
    "    '--input-recursive INPUT_DIR1': [f\"{bucket}/{gwas_dir}/imported/step3_docker/step3_docker/WES470k_{maf[i]}_{anno[i]}_chr{chrom[i]}/\" for i in range(tot_len)],\n",
    "    '--input-recursive INPUT_DIR2': [f\"{bucket}/{gwas_dir}/step3_output/{maf[i]}_{anno[i]}_chr{chrom[i]}/\" for i in range(tot_len)],\n",
    "    '--input INPUT_GWAS1': [f\"{bucket}/{gwas_dir}/imported/step2_phenome/step2/WES470k_Whites_{phecodes_all[i]}_chr{chrom[i]}_gene\" for i in range(tot_len)],\n",
    "    '--input INPUT_GWAS2': [f\"{bucket}/{gwas_dir}/step2_output/Pheno_{phecodes_all[i]}_chr{chrom[i]}_step2_output_single.txt\" for i in range(tot_len)],\n",
    "    '--env maf': maf,\n",
    "    '--env Phecode': phecodes_all,\n",
    "    '--output-recursive OUT_DIR': [f\"{bucket}/{gwas_dir}/step4_output/\" for _ in range(tot_len)],\n",
    "    '--env anno': anno,\n",
    "    '--env chrom': chrom,    \n",
    "})\n",
    "PARAMETER_FILENAME = f'{JOB_NAME}_params.tsv'\n",
    "%env PARAMETER_FILENAME={PARAMETER_FILENAME}\n",
    "%env IMAGE_PATH={} # Please supply\n",
    "params_df.to_csv(PARAMETER_FILENAME, sep='\\t', index=False)\n",
    "\n",
    "job_id = !source ~/aou_dsub.bash; aou_dsub \\\n",
    "  --name \"${JOB_NAME}\" \\\n",
    "  --provider google-batch \\\n",
    "  --regions us-central1 \\\n",
    "  --image \"${IMAGE_PATH}/meta-saige:0.3.0\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/dsub_logs/saige_step4\" \\\n",
    "  --boot-disk-size 30 \\\n",
    "  --disk-size 50 \\\n",
    "  --min-ram 10 \\\n",
    "  --tasks \"${PARAMETER_FILENAME}\" \\\n",
    "  --script \"Cmd_meta.sh\"\n",
    "#print(\"\\n\".join(job_id))\n",
    "job_id = job_id[1].split(\" \")[-1]\n",
    "%env JOB_ID={job_id}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "## Multi-Ancestry, Meta-Analysis with META-SAIGE using AllofUS (EUR, AFR, AMR) and UKB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd_line = '''\n",
    "INPUT_DIR1=${BUCKET}/SAIGE_GENE/imported/step3_docker/step3_docker/WES470k_${maf}_${anno}_chr${i}/\n",
    "INPUT_DIR2=${BUCKET}/SAIGE_GENE/step3_output/${maf}_${anno}_chr${i}/\n",
    "INPUT_DIR3=${BUCKET}/SAIGE_GENE/step3_output/afr_${maf}_${anno}_chr${i}/\n",
    "INPUT_DIR4=${BUCKET}/SAIGE_GENE/step3_output/amr_${maf}_${anno}_chr${i}/\n",
    "INPUT_GWAS1=${BUCKET}/SAIGE_GENE/imported/step2_phenome/step2/WES470k_Whites_${Phecode}_chr${i}_gene\n",
    "INPUT_GWAS2=${BUCKET}/SAIGE_GENE/step2_output/Pheno_${Phecode}_chr${i}_step2_output_single.txt\n",
    "INPUT_GWAS3=${BUCKET}/SAIGE_GENE/step2_output/${phen}_afr_chr${i}_step2_output_single.txt\n",
    "INPUT_GWAS4=${BUCKET}/SAIGE_GENE/step2_output/${phen}_amr_chr${i}_step2_output_single.txt\n",
    "head -1 $INPUT_GWAS1 > /app/tmp_chr${i}_GWAS1.txt ; tail -n+2 $INPUT_GWAS1 | awk 'BEGIN{OFS=\"\\\\t\"}$1=\"chr\"$1, $3=\"chr\"$3'  >> /app/tmp_chr${i}_GWAS1.txt \n",
    "head -1 $INPUT_DIR1/WES470k_chr${i}.marker_info.txt > /app/tmp_chr${i}_mkr_info1.txt ; tail -n+2 $INPUT_DIR1/WES470k_chr${i}.marker_info.txt | awk 'BEGIN{OFS=\"\\\\t\"}$1=\"chr\"$1' >> /app/tmp_chr${i}_mkr_info1.txt \n",
    "cp ${BUCKET}/SAIGE_GENE/saige_meta_scripts/Lib_GC.R /app/\n",
    "cp ${BUCKET}/SAIGE_GENE/saige_meta_scripts/MetaSAIGE.R /app/\n",
    "chrom=$i\n",
    "/usr/bin/time --verbose Rscript /app/RV_meta_GC.R \\\n",
    "    --num_cohorts 4 \\\n",
    "    --chr ${chrom} \\\n",
    "    --col_co 10 \\\n",
    "    --info_file_path $INPUT_DIR1/WES470k_chr${chrom}.marker_info.txt $INPUT_DIR2/${maf}_${anno}_chr${chrom}_loftee.marker_info.txt $INPUT_DIR3/afr_${maf}_${anno}_chr${chrom}_loftee.marker_info.txt $INPUT_DIR4/amr_${maf}_${anno}_chr${chrom}_loftee.marker_info.txt \\\n",
    "    --gene_file_prefix $INPUT_DIR1/WES470k_chr${chrom}_ $INPUT_DIR2/${maf}_${anno}_chr${chrom}_loftee_ $INPUT_DIR3/afr_${maf}_${anno}_chr${chrom}_loftee_ $INPUT_DIR4/amr_${maf}_${anno}_chr${chrom}_loftee_ \\\n",
    "    --gwas_path $INPUT_GWAS1 $INPUT_GWAS2 $INPUT_GWAS3 $INPUT_GWAS4 \\\n",
    "    --ancestry 1 1 2 3 \\\n",
    "    --trait_type binary \\\n",
    "    --verbose FALSE \\\n",
    "    --output_prefix $OUT_DIR/Multi_ancestry_${phen}_${maf}_${anno}_chr${chrom}_GCfix\n",
    "'''\n",
    "with open(\"Cmd_ma_recoded.sh\", \"w\") as text_file:\n",
    "    text_file.write(cmd_line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')\n",
    "%env USER_NAME={USER_NAME}\n",
    "bucket = os.getenv('WORKSPACE_BUCKET')\n",
    "gwas_dir = 'SAIGE_GENE'\n",
    "maf=(['0.01']*22+['0.001']*22+['0.0001']*22)*3*2\n",
    "anno=(['lof']*66+['missense_lof']*66+['missense_lof_synonymous']*66)*2\n",
    "phen = ['T2D']*198+['colorectal']*198\n",
    "phen_ukb = ['t2d']*198+['colca']*198\n",
    "phecodes = ['250.2']*198+['153']*198\n",
    "chrom=[i for i in range(1,23)]*3*3*2\n",
    "tot_len = len(maf)\n",
    "params_df0 = pd.DataFrame(data={\n",
    "    '--env maf': maf,\n",
    "    '--env phen': phen,\n",
    "    '--env phen_ukb': phen_ukb,\n",
    "    '--output-recursive OUT_DIR': [f\"{bucket}/{gwas_dir}/step4_output_recode/\" for _ in range(tot_len)],\n",
    "    '--env Phecode': phecodes,\n",
    "    '--env anno': anno,\n",
    "    '--env i': chrom,    \n",
    "})\n",
    "params_df0=params_df0.loc[(params_df0['--env i'].isin([2]))]\n",
    "params_df0.to_csv('saige-step4_ma.tsv',sep = '\\t', index = False)\n",
    "%env PARAMETER_FILENAME=saige-step4_ma.tsv\n",
    "%env JOB_NAME=meta_mult\n",
    "job_id = !source ~/aou_dsub.bash; aou_dsub \\\n",
    "  --name \"${JOB_NAME}\" \\\n",
    "  --provider google-cls-v2 \\\n",
    "  --image \"gcr.io/pheweb/meta-saige:0.2.3\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/dsub_logs/saige_step4\" \\\n",
    "  --boot-disk-size 10 \\\n",
    "  --disk-size 20 \\\n",
    "  --min-ram 15 \\\n",
    "  --tasks \"${PARAMETER_FILENAME}\" \\\n",
    "  --mount BUCKET=\"${WORKSPACE_BUCKET}\" \\\n",
    "  --script \"Cmd_ma_recoded.sh\"\n",
    "#print(\"\\n\".join(job_id))\n",
    "job_id2 = job_id[1].split(\" \")[-1]\n",
    "%env JOB_ID={job_id2}\n",
    "job_id"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
