{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fd15c36d",
   "metadata": {},
   "source": [
    "# Phenotype Generation for Power Evaluation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66d78f99",
   "metadata": {},
   "source": [
    "### Define functions for generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a39a180",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# Load packages\n",
    "\n",
    "if (!require(seqminer)) install.packages(seqminer)\n",
    "if (!require(data.table)) install.packages(data.table)\n",
    "if (!require(tibble)) install.packages(tibble)\n",
    "if (!require(dplyr)) install.packages(dplyr)\n",
    "if (!require(Matrix)) install.packages(Matrix)\n",
    "if (!require(sparsesvd)) install.packages(sparsesvd)\n",
    "\n",
    "library(seqminer)\n",
    "library(data.table)\n",
    "library(tibble)\n",
    "library(dplyr)\n",
    "library(Matrix)\n",
    "library(sparsesvd)\n",
    "\n",
    "# Load PLINK file\n",
    "load_plink <- function(plink_prefix) {\n",
    "    plink_obj <- seqminer::openPlink(plink_prefix)\n",
    "    marker_index <- seq(nrow(plink_obj$bim))\n",
    "    sample_index <- seq(nrow(plink_obj$fam))\n",
    "    plink_matrix <- seqminer::readPlinkToMatrixByIndex(plink_prefix, sample_index, marker_index)\n",
    "    colnames(plink_matrix) <- plink_obj$bim$V2\n",
    "\n",
    "    return(plink_matrix)\n",
    "}\n",
    "\n",
    "# Read Group file and split variants by functional annotations\n",
    "read_groupfile <- function(groupfile_name, gene_name) {\n",
    "    groupfile <- file(groupfile_name, \"r\")\n",
    "    line <- 0\n",
    "    var <- NULL\n",
    "    anno <- NULL\n",
    "\n",
    "    while (TRUE) {\n",
    "        line <- line + 1\n",
    "        marker_group_line <- readLines(groupfile, n = 1)\n",
    "\n",
    "        if (length(marker_group_line) == 0) {\n",
    "            break\n",
    "        }\n",
    "\n",
    "        marker_group_line_list <- strsplit(marker_group_line, split = c(\" +\", \"\\t\"))[[1]]\n",
    "\n",
    "        if (marker_group_line_list[1] == gene_name) {\n",
    "            if (marker_group_line_list[2] == \"var\") {\n",
    "                var <- marker_group_line_list\n",
    "            } else {\n",
    "                anno <- marker_group_line_list\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    lof_idx <- which(anno == \"lof\")\n",
    "    mis_idx <- which(anno == \"missense\")\n",
    "    syn_idx <- which(anno == \"synonymous\")\n",
    "\n",
    "    lof_var <- var[lof_idx]\n",
    "    mis_var <- var[mis_idx]\n",
    "    syn_var <- var[syn_idx]\n",
    "\n",
    "    out <- list(lof_var, mis_var, syn_var)\n",
    "    return(out)\n",
    "}\n",
    "\n",
    "\n",
    "# Convert genotypes for missing / alt-first PLINK files\n",
    "convert_missing <- function(plink_matrix, ref_first=FALSE) {\n",
    "    if (ref_first == FALSE) {   # alt-first\n",
    "        # Convert missing genotypes to reference-homozygotes\n",
    "        plink_matrix[which(plink_matrix == -9)] <- 2\n",
    "        # Flip genotypes\n",
    "        plink_matrix <- 2 - plink_matrix\n",
    "    } else {\n",
    "        plink_matrix[which(plink_matrix == -9)] <- 0\n",
    "    }\n",
    "\n",
    "    return(plink_matrix)\n",
    "}\n",
    "\n",
    "\n",
    "# function for counting MAF of input genotype matrix (used in split_plink_matrix)\n",
    "count_maf <- function(geno_mat) {\n",
    "    mac_vec <- rep(0, ncol(geno_mat))\n",
    "    for (i in seq_len(ncol(geno_mat))) {\n",
    "        # n0 <- length(which(geno_mat[, i]) == 0)\n",
    "        n1 <- length(which(geno_mat[, i] == 1))\n",
    "        n2 <- length(which(geno_mat[, i] == 2))\n",
    "\n",
    "        mac <- n1 + 2 * n2\n",
    "        mac_vec[i] <- mac\n",
    "    }\n",
    "    maf_vec <- mac_vec / (2 * nrow(geno_mat))\n",
    "    common_list <- which(maf_vec > 0.01)\n",
    "\n",
    "    out <- list(mac_vec, common_list)\n",
    "    return(out)\n",
    "}\n",
    "\n",
    "\n",
    "# Split genotype matrix by function annotation (lof / mis / syn)\n",
    "split_plink_matrix <- function(plink_matrix_converted, lof_var, mis_var, syn_var, mac_threshold = 10) {\n",
    "    lof_mat <- plink_matrix_converted[, var_by_func_anno[[1]], drop = F]\n",
    "    mis_mat <- plink_matrix_converted[, var_by_func_anno[[2]], drop = F]\n",
    "    syn_mat <- plink_matrix_converted[, var_by_func_anno[[3]], drop = F]\n",
    "\n",
    "    # count MAF for each functional annotation\n",
    "    mac_lof <- count_maf(lof_mat)[[1]]\n",
    "    common_list_lof <- count_maf(lof_mat)[[2]]\n",
    "\n",
    "    mac_mis <- count_maf(mis_mat)[[1]]\n",
    "    common_list_mis <- count_maf(mis_mat)[[2]]\n",
    "\n",
    "    mac_syn <- count_maf(syn_mat)[[1]]\n",
    "    common_list_syn <- count_maf(syn_mat)[[2]]\n",
    "\n",
    "    # collapsing columns with MAC < threshold\n",
    "    collapse_lof <- which(mac_lof < mac_threshold)\n",
    "    collapse_mis <- which(mac_mis < mac_threshold)\n",
    "    collapse_syn <- which(mac_syn < mac_threshold)\n",
    "\n",
    "    if(length(collapse_lof) > 0) {\n",
    "        rowsum_lof <- rowSums(lof_mat[, collapse_lof, drop = F])\n",
    "        rowsum_lof[which(rowsum_lof > 1)] <- 1\n",
    "        lof_mat_collapsed <- cbind(lof_mat[, -c(collapse_lof, common_list_lof), drop = F], rowsum_lof)\n",
    "    } else if ((length(common_list_lof) > 0)) {\n",
    "        lof_mat_collapsed <- lof_mat[, -c(common_list_lof), drop = F]\n",
    "    } else {\n",
    "        lof_mat_collapsed <- lof_mat\n",
    "    }\n",
    "\n",
    "    if(length(collapse_mis) > 0) {\n",
    "        rowsum_mis <- rowSums(mis_mat[, collapse_mis, drop = F])\n",
    "        rowsum_mis[which(rowsum_mis > 1)] <- 1\n",
    "        mis_mat_collapsed <- cbind(mis_mat[, -c(collapse_mis, common_list_mis), drop = F], rowsum_mis)\n",
    "    } else if ((length(common_list_mis) > 0)) {\n",
    "        mis_mat_collapsed <- mis_mat[, -c(common_list_mis), drop = F]\n",
    "    } else {\n",
    "        mis_mat_collapsed <- mis_mat\n",
    "    }\n",
    "\n",
    "    if(length(collapse_syn) > 0) {\n",
    "        rowsum_syn <- rowSums(syn_mat[, collapse_syn, drop = F])\n",
    "        rowsum_syn[which(rowsum_syn > 1)] <- 1\n",
    "        syn_mat_collapsed <- cbind(syn_mat[, -c(collapse_syn, common_list_syn), drop = F], rowsum_syn)\n",
    "    } else if ((length(common_list_syn) > 0)) {\n",
    "        syn_mat_collapsed <- syn_mat[, -c(common_list_syn), drop = F]\n",
    "    } else {\n",
    "        syn_mat_collapsed <- syn_mat\n",
    "    }\n",
    "\n",
    "    out <- list(lof_mat_collapsed, mis_mat_collapsed, syn_mat_collapsed)\n",
    "    return(out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37ef3c26",
   "metadata": {},
   "source": [
    "### Phenotype by common variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48148b26",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# Install packages\n",
    "\n",
    "if (!require(seqminer)) install.packages(seqminer)\n",
    "if (!require(data.table)) install.packages(data.table)\n",
    "\n",
    "# Load libraries\n",
    "\n",
    "setwd(\"/media/leelabsg-storage0/kisung/META-SAIGE/script\")\n",
    "library(seqminer)\n",
    "library(data.table)\n",
    "source(\"230710_functions.R\")\n",
    "\n",
    "plink_prefix <- paste0(\"/media/leelabsg-storage0/kisung/META-SAIGE/data/genotype/common/WES_common_merged\")\n",
    "plink_matrix <- load_plink(plink_prefix)\n",
    "plink_matrix_converted <- convert_missing(plink_matrix, ref_first = FALSE)\n",
    "print(\"PLINK file for common variants loaded\")\n",
    "\n",
    "# Generate covariates and betas\n",
    "set.seed(0)\n",
    "n_sim <- 100\n",
    "n_sample <- nrow(plink_matrix_converted)\n",
    "\n",
    "mean_vec <- colMeans(plink_matrix_converted)\n",
    "plink_matrix_centered <- plink_matrix_converted - rep(mean_vec, rep.int(nrow(plink_matrix_converted), ncol(plink_matrix_converted)))\n",
    "\n",
    "for (n in 1:n_sim) {\n",
    "    n_common <- ncol(plink_matrix_centered)\n",
    "    cov1 <- rbinom(n_sample, 1, 0.1)\n",
    "    cov2 <- rnorm(n_sample, 0, 1)\n",
    "    betas <- rnorm(n_common, 0, 1 / n_common)\n",
    "\n",
    "    # Phenotype generation (Common variant part)\n",
    "    pheno_common <- plink_matrix_centered %*% betas\n",
    "\n",
    "    out <- cbind(rownames(pheno_common), pheno_common, cov1, cov2)\n",
    "    colnames(out)[1] <- \"IID\"\n",
    "    colnames(out)[2] <- \"pheno_common\"\n",
    "    outname <- paste0(\"/media/leelabsg-storage0/kisung/META-SAIGE/data/phenotype/common_231005/pheno_sim\", n, \".txt\")\n",
    "\n",
    "    write.table(out, outname, row.names=F, col.names=T, quote=F)\n",
    "    print(paste(\"Simulation\", n, \"completed\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4281a3c",
   "metadata": {},
   "source": [
    "### Phenotype by rare variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1907702b",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# Install packages\n",
    "\n",
    "if (!require(seqminer)) install.packages(seqminer)\n",
    "if (!require(data.table)) install.packages(data.table)\n",
    "\n",
    "# Load libraries\n",
    "\n",
    "setwd(\"/media/leelabsg-storage0/kisung/META-SAIGE/script\")\n",
    "library(seqminer)\n",
    "library(data.table)\n",
    "source(\"230710_functions.R\")\n",
    "\n",
    "# Generate covariates and betas\n",
    "set.seed(0)\n",
    "n_sim <- 100\n",
    "groupfile_name <- \"/media/leelabsg-storage0/kisung/dnanexus/group_files/UKBexome_all_chr.txt\"\n",
    "gene_list <- c(\"IGSF9B\", \"BRCA2\", \"APOB\", \"GPRIN1\", \"CFB\", \"DDR1\", \"GPSM3\", \"HLA-DRB1\", \"DBH\", \"IL33\")\n",
    "chr_list <- c(11, 13, 2, 5, 6, 6, 6, 6, 9, 9)\n",
    "\n",
    "for (i in seq_len(length(gene_list))) {\n",
    "    gene_name <- as.character(gene_list[i])\n",
    "    chr <- chr_list[i]\n",
    "\n",
    "    rare_plink_prefix <- paste0(\"/media/leelabsg-storage0/kisung/META-SAIGE/data/genotype/rare_231005/\", chr, \"_\", gene_name)\n",
    "    rare_plink_matrix <- load_plink(rare_plink_prefix)\n",
    "    rare_plink_matrix_converted <- convert_missing(rare_plink_matrix, ref_first = FALSE)\n",
    "    n_sample <- nrow(rare_plink_matrix_converted)\n",
    "\n",
    "    # mat_by_func_anno = [G_lof, G_mis, G_syn]\n",
    "    var_by_func_anno <- read_groupfile(groupfile_name, gene_name)\n",
    "    mat_by_func_anno <- split_plink_matrix(rare_plink_matrix_converted,\n",
    "        var_by_func_anno[[1]],\n",
    "        var_by_func_anno[[2]],\n",
    "        var_by_func_anno[[3]],\n",
    "        mac_threshold = 1\n",
    "    )\n",
    "\n",
    "    # rare_plink_matrix_converted[, var_by_func_anno[[2]]]\n",
    "\n",
    "    # Genotype matrix for rare_variants\n",
    "    lof_mat <- mat_by_func_anno[[1]][, c(1:(ncol(mat_by_func_anno[[1]]) - 1))]\n",
    "    mis_mat <- mat_by_func_anno[[2]][, c(1:(ncol(mat_by_func_anno[[2]]) - 1))]\n",
    "    syn_mat <- mat_by_func_anno[[3]][, c(1:(ncol(mat_by_func_anno[[3]]) - 1))]\n",
    "\n",
    "    # Proportion of causal variants\n",
    "    ## Scenario : (LoF, mis, syn) = (20%, 10%, 2%), (30%, 10%, 2%)\n",
    "    lof_causal_rates <- c(0.2, 0.3)\n",
    "    mis_causal_rates <- c(0.1, 0.1)\n",
    "    syn_causal_rates <- c(0.02, 0.02)\n",
    "\n",
    "    # Absolute effect sizes\n",
    "    ## Scenarios : (LoF, mis, syn) = (0.3 log MAF, 0.15 log MAF, 0.15 log MAF), (0.5 log MAF, 0.25 log MAF, 0.25 log MAF)\n",
    "    lof_effect_sizes <- c(0.3, 0.5, 0.7)\n",
    "    mis_effect_sizes <- c(0.15, 0.25, 0.35)\n",
    "    syn_effect_sizes <- c(0.15, 0.25, 0.35)\n",
    "\n",
    "    # Effect directions\n",
    "    ## Scenarios : (LoF, mis, syn) = (100%, 80%, 50%), (100%, 100%, 100%)\n",
    "    lof_direction <- c(1, 1)\n",
    "    mis_direction <- c(0.8, 1)\n",
    "    syn_direction <- c(0.5, 1)\n",
    "\n",
    "    # MAC by functional annotation\n",
    "    lof_MAC <- colSums(lof_mat)\n",
    "    mis_MAC <- colSums(mis_mat)\n",
    "    syn_MAC <- colSums(syn_mat)\n",
    "\n",
    "    # Generate causal vector\n",
    "    betas_lof <- rep(0, ncol(lof_mat))\n",
    "    betas_mis <- rep(0, ncol(mis_mat))\n",
    "    betas_syn <- rep(0, ncol(syn_mat))\n",
    "\n",
    "    for (n in 1:n_sim) {\n",
    "        for (ii in 1:2) {\n",
    "            for (jj in 1:3) {\n",
    "                for (kk in 1:2) {\n",
    "                    # Generate beta for LoF\n",
    "                    for (j in seq_len(ncol(lof_mat))) {\n",
    "                        MAF <- lof_MAC[j] * 2 / n_sample\n",
    "                        if (lof_MAC[j] <= 10) {\n",
    "                            betas_lof[j] <- rbinom(1, 1, 3 * lof_causal_rates[ii]) * abs(lof_effect_sizes[jj] * log(MAF, base = 10))\n",
    "                        } else {\n",
    "                            betas_lof[j] <- rbinom(1, 1, lof_causal_rates[ii]) * abs(lof_effect_sizes[jj] * log(MAF, base = 10))\n",
    "                        }\n",
    "                    }\n",
    "\n",
    "                    # Generate beta for mis\n",
    "                    for (j in seq_len(ncol(mis_mat))) {\n",
    "                        MAF <- mis_MAC[j] * 2 / n_sample\n",
    "                        if (mis_MAC[j] <= 10) {\n",
    "                            betas_mis[j] <- rbinom(1, 1, 3 * mis_causal_rates[ii]) * abs(mis_effect_sizes[jj] * log(MAF, base = 10)) * (2 * (rbinom(1, 1, mis_direction[kk]) - 0.5))\n",
    "                        } else {\n",
    "                            betas_mis[j] <- rbinom(1, 1, mis_causal_rates[ii]) * abs(mis_effect_sizes[jj] * log(MAF, base = 10)) * (2 * (rbinom(1, 1, mis_direction[kk]) - 0.5))\n",
    "                        }\n",
    "                    }\n",
    "\n",
    "                    # Generate beta for syn\n",
    "                    for (j in seq_len(ncol(syn_mat))) {\n",
    "                        MAF <- syn_MAC[j] * 2 / n_sample\n",
    "                        if (syn_MAC[j] <= 10) {\n",
    "                            betas_syn[j] <- rbinom(1, 1, 3 * syn_causal_rates[ii]) * abs(syn_effect_sizes[jj] * log(MAF, base = 10)) * (2 * (rbinom(1, 1, syn_direction[kk]) - 0.5))\n",
    "                        } else {\n",
    "                            betas_syn[j] <- rbinom(1, 1, syn_causal_rates[ii]) * abs(syn_effect_sizes[jj] * log(MAF, base = 10)) * (2 * (rbinom(1, 1, syn_direction[kk]) - 0.5))\n",
    "                        }\n",
    "                    }\n",
    "\n",
    "                    # Phenotype generation (rare-variant part)\n",
    "                    pheno_lof <- lof_mat %*% betas_lof\n",
    "                    pheno_mis <- mis_mat %*% betas_mis\n",
    "                    pheno_syn <- syn_mat %*% betas_syn\n",
    "\n",
    "                    pheno_rv <- pheno_lof + pheno_mis + pheno_syn\n",
    "\n",
    "                    # betas_lof_name <- paste0(\"/media/leelabsg-storage0/kisung/META-SAIGE/data/phenotype/rare/betas_lof_sim\", n, \"_gene\", i, \"_\", ii, \"_\", jj, \"_\", kk, \".txt\")\n",
    "                    # betas_mis_name <- paste0(\"/media/leelabsg-storage0/kisung/META-SAIGE/data/phenotype/rare/betas_mis_sim\", n, \"_gene\", i, \"_\", ii, \"_\", jj, \"_\", kk, \".txt\")\n",
    "                    # betas_syn_name <- paste0(\"/media/leelabsg-storage0/kisung/META-SAIGE/data/phenotype/rare/betas_syn_sim\", n, \"_gene\", i, \"_\", ii, \"_\", jj, \"_\", kk, \".txt\")\n",
    "\n",
    "                    out <- cbind(rownames(pheno_rv), pheno_rv)\n",
    "                    colnames(out) <- c(\"IID\", \"pheno_rv\")\n",
    "\n",
    "                    outname <- paste0(\"/media/leelabsg-storage0/kisung/META-SAIGE/data/phenotype/rare_231005/pheno_rare_sim\", n, \"_gene\", i, \"_\", ii, \"_\", jj, \"_\", kk, \".txt\")\n",
    "\n",
    "                    # write.table(betas_lof, betas_lof_name, row.names=F, col.names=F, quote=F)\n",
    "                    # write.table(betas_mis, betas_mis_name, row.names=F, col.names=F, quote=F)\n",
    "                    # write.table(betas_syn, betas_syn_name, row.names=F, col.names=F, quote=F)\n",
    "                    write.table(out, outname, row.names=F, col.names=T, quote=F)\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    print(paste0(\"Gene \", i, \" completed\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c12cc445",
   "metadata": {},
   "source": [
    "### Make binary phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bce85cf4",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(dplyr)\n",
    "\n",
    "setwd(\"/media/leelabsg-storage0/kisung/META-SAIGE/data/phenotype\")\n",
    "\n",
    "n_sim <- 100\n",
    "\n",
    "f <- function(x, pheno, p) {\n",
    "    mean(1 / (1 + exp(-(pheno + x)))) - p\n",
    "}\n",
    "\n",
    "for (n in 1:n_sim) {\n",
    "    pheno_common_name <- paste0(\"common_231005/pheno_sim\", n, \".txt\")\n",
    "    pheno_common <- fread(pheno_common_name)\n",
    "    for (ii in 1:2) {\n",
    "        for (jj in 1:3) {\n",
    "            for (kk in 1:2) {\n",
    "                for (j in 1:10) {\n",
    "                    pheno_rare_name <- paste0(\"rare_231005/pheno_rare_sim\", n, \"_gene\", j, \"_\", ii, \"_\", jj, \"_\", kk, \".txt\")\n",
    "                    pheno_rare <- fread(pheno_rare_name)\n",
    "\n",
    "                    if (j == 1) {\n",
    "                        pheno_rare_total <- pheno_rare\n",
    "                    } else {\n",
    "                        pheno_rare_total$pheno_rv <- pheno_rare_total$pheno_rv + pheno_rare$pheno_rv\n",
    "                    }\n",
    "                }\n",
    "                pheno <- left_join(pheno_common, pheno_rare_total, by=\"IID\")\n",
    "                pheno$pheno_total <- pheno$pheno_common + pheno$cov1 + pheno$cov2 + pheno$pheno_rv\n",
    "\n",
    "                #prev <- c(0.01, 0.05, 0.1, 0.2)\n",
    "                prev <- c(0.01, 0.05)\n",
    "                for (pr in prev) {\n",
    "                    a <- uniroot(f, c(-100, 100), pheno = pheno$pheno_total, p = pr)$root\n",
    "                    prob <- 1 / (1 + exp(-(pheno$pheno_total + a)))\n",
    "\n",
    "                    pheno_binary = rbinom(nrow(pheno), 1, prob)\n",
    "                    pheno_out <- cbind(pheno, pheno_binary)\n",
    "                    outname <- paste0(\"./total_231005/pheno_sim\", n, \"_\", ii, \"_\", jj, \"_\", kk, \"_prev\", pr, \".txt\")\n",
    "\n",
    "                    write.table(pheno_out, outname, row.names=F, col.names=T, quote=F)\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    print(paste0(\"Data for simulation \", n, \" generated.\"))\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
